src/rust/protocol/src/memcache/wire/request/parse.rs

```diff
@@ -0,0 +1,368 @@
+            .ok_or(ParseError::Invalid)?;
+
+        let bytes_end = if let Some(next_space) = next_space {
+            // if we have both, bytes_end is before the earlier of the two
+            if next_space < first_crlf {
+                // validate that noreply isn't malformed
+                if &buffer[(next_space + 1)..(first_crlf)] == NOREPLY.as_bytes() {
+                    noreply = true;
+                    next_space
+                } else {
+                    return Err(ParseError::Invalid);
+                }
+            } else {
+                first_crlf
+            }
+        } else {
+            first_crlf
+        };
+
+        // this checks for malformed requests where a CRLF is at an
+        // unexpected part of the request
+        if (expiry_end + 1) >= bytes_end {
+            return Err(ParseError::Invalid);
+        }
+
+        if let Ok(Ok(bytes)) =
+            std::str::from_utf8(&buffer[(expiry_end + 1)..bytes_end]).map(|v| v.parse::<usize>())
+        {
+            let consumed = first_crlf + CRLF.len() + bytes + CRLF.len();
+            if buffer.len() >= consumed {
+                let key = buffer[(cmd_end + 1)..key_end].to_vec().into_boxed_slice();
+                let value = buffer[(first_crlf + CRLF.len())..(first_crlf + CRLF.len() + bytes)]
+                    .to_vec()
+                    .into_boxed_slice();
+
+                let entry = MemcacheEntry {
+                    key,
+                    value,
+                    cas: None,
+                    expiry,
+                    flags,
+                };
+                Ok(ParseOk {
+                    message: MemcacheRequest::Set { entry, noreply },
+                    consumed,
+                })
+            } else {
+                // the buffer doesn't yet have all the bytes for the value
+                Err(ParseError::Incomplete)
+            }
+        } else {
+            // expiry couldn't be parsed
+            Err(ParseError::Invalid)
+        }
+    } else {
+        // no space (' ') in the buffer
+        Err(ParseError::Incomplete)
+    }
+}
+
+fn parse_add(buffer: &[u8]) -> Result<ParseOk<MemcacheRequest>, ParseError> {
+    let request = parse_set(buffer)?;
+    let consumed = request.consumed();
+
+    let message = if let MemcacheRequest::Set { entry, noreply } = request.into_inner() {
+        MemcacheRequest::Add { entry, noreply }
+    } else {
+        unreachable!()
+    };
+
+    Ok(ParseOk { message, consumed })
+}
+
+fn parse_replace(buffer: &[u8]) -> Result<ParseOk<MemcacheRequest>, ParseError> {
+    let request = parse_set(buffer)?;
+    let consumed = request.consumed();
+
+    let message = if let MemcacheRequest::Set { entry, noreply } = request.into_inner() {
+        MemcacheRequest::Replace { entry, noreply }
+    } else {
+        unreachable!()
+    };
+
+    Ok(ParseOk { message, consumed })
+}
+
+fn parse_cas(buffer: &[u8]) -> Result<ParseOk<MemcacheRequest>, ParseError> {
```

Comment:
Not a huge fan of the code duplication between `cas` and other write requests. All write requests have exactly one verb and one key (this includes `incr`/`decr`/`append`/`prepend` which are not implemented right now, so it could get worse :P ) , and between `set`/`add`/`replace`/`cas` the only difference in format is `cas` has an additional `cas_value` field. I think it's possible to consolidate the parser logic further (I know it's a bit of a pain, I rewrote my version at least 3 times).

Reply:
Should be better now.